velox/common/caching/GroupTracker.h

```diff
@@ -0,0 +1,208 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include "velox/common/base/BloomFilter.h"
+#include "velox/common/caching/FileIds.h"
+#include "velox/common/caching/ScanTracker.h"
+#include "velox/common/caching/StringIdMap.h"
+
+#include <folly/container/F14Map.h>
+#include <folly/container/F14Set.h>
+
+namespace facebook::velox::cache {
+
+class ApproxCounter {
+ public:
+  ApproxCounter(int32_t range) {
+    bloom_.reset(range);
+  }
+
+  void add(uint64_t value) {
+    if (!bloom_.mayContain(value)) {
+      bloom_.insert(value);
+      ++count_;
+    }
+  }
+
+  int32_t count() const {
+    return count_;
+  }
+
+ private:
+  BloomFilter<true> bloom_;
+  int32_t count_{0};
+};
+
+// Represents a groupId, column and its size and score. These are
+// sorted and as many are selected from the top as will fit on SSD.
+struct SsdScore {
+  // Number used for ranking. Correlates to access frequency and
+  // inversely to size.
+  float score;
+
+  // Expected size in bytes for caching to SSD
+  float size;
+  // Recorded read activity, with older reads decayed.
+  float readBytes;
+  uint64_t groupId;
+  int32_t columnId;
+};
+
+// Tracks usage of different columns of inside a file group. Tracks the number
+// of distinct files in the group and their average number of splits. Produces
+// estimates of the column size for each accessed column in the group.
+class GroupTracker {
+ public:
+  static constexpr int32_t kExpectedNumFiles = 100;
+
+  // Constructs a tracker for file group 'ame'. 'name' is for example
+  // the directory path of a Hive partition.
+  GroupTracker(const StringIdLease& name)
+      : name_(name), numFiles_(kExpectedNumFiles) {}
+
+  // Records that 'fileId' belongs to this group and has 'numStripes'
+  // stripes. Used to scale up the column sizes reported by
+  // recordReference().
+  void recordFile(uint64_t fileId, int32_t numStripes);
+
+  // Records a sample stripe size for 'columnId'.
+  void recordReference(uint64_t fileId, int32_t columnId, int32_t bytes);
+
+  // Records read of 'bytes' from 'columnId' This is compared to the
+  // reference size to get read density.
+  void recordRead(uint64_t fileId, int32_t columnId, int32_t bytes);
+
+  // Adds the column scores to 'scores'. If 'decayPct' is non-0,
+  // decays the recorded accesses by 'decayPct'% but at least by one
+  // whole access.
+  void addColumnScores(int32_t decayPct, std::vector<SsdScore>& scores);
+
+  bool eraseColumn(int32_t columnId) {
+    columns_.erase(columnId);
+    return columns_.empty();
+  }
+
+ private:
+  StringIdLease name_;
+
+  // Map of column to access data.
+  folly::F14FastMap<int32_t, TrackingData> columns_;
+
+  // Count of distinct files seen in recordFile().
+  ApproxCounter numFiles_;
```

Comment:
I'm seeing that numFiles_ is initialized with 100 which is used as a bloom filter capacity. What happens if there are > 100 files?